src/browser/modules/Editor/ProjectFilesButton.tsx

```diff
@@ -0,0 +1,205 @@
+import floppyDisk from 'icons/floppy-disk.svg'
+import * as drawer from 'shared/modules/sidebar/sidebarDuck'
+import { EDIT_CONTENT, SET_CONTENT } from 'shared/modules/editor/editorDuck'
+import {
+  ADD_PROJECT_FILE,
+  SELECT_PROJECT_FILE,
+  IProjectFile,
+  SAVE_PROJECT_FILE,
+  PROJECT_FILE_ERROR,
+  EDIT_PROJECT_FILE_START,
+  EDIT_PROJECT_FILE_END,
+  PROJECT_FILES_MOUNTED,
+  PROJECT_FILES_UNMOUNTED
+} from 'browser/modules/Sidebar/project-files.constants'
+import { useMutation } from '@apollo/client'
+
+const IS_RELATE_AVAILABLE = true // @todo: will come from Redux store eventually
+const PROJECT_EDITOR_BUTTON_TITLE = 'Project File'
+const PROJECT_SAVE_EDITOR_BUTTON_TITLE = 'Save Project File'
+const PROJECT_EDIT_EDITOR_BUTTON_TITLE = 'Edit Project File'
+const PROJECT_FILES_DRAWER_NAME = 'project files'
+
+interface ProjectFilesButtonProps {
+  width?: number
+  editorValue: () => string
+  bus: Bus
+  toggleDrawer: () => void
+  isDrawerOpen: boolean
+}
+
+type IActiveRelateFile = Omit<IProjectFile, 'downloadToken'>
+
+const ProjectFileButton = ({
+  width = 24,
+  editorValue,
+  bus,
+  toggleDrawer,
+  isDrawerOpen
+}: ProjectFilesButtonProps): JSX.Element | null => {
+  const [addFile, { error }] = useMutation(ADD_PROJECT_FILE)
+  const [isEditMode, setEditMode] = useState(false)
+  const [isSaveModeClick, setSaveModeClick] = useState(false)
+  const [isProjectFilesMounted, setProjectFilesMounted] = useState(false)
+  const [
+    activeRelateFile,
+    setActiveRelateFile
+  ] = useState<IActiveRelateFile | null>(null)
+
+  useEffect(() => {
+    let isStillMounted = true
+
+    // when a saved Project Script or Local Cache Script is clicked
+    // not sure at this point which it could be
+    bus &&
+      bus.take(EDIT_CONTENT, () => {
+        if (isStillMounted) {
+          setEditMode(false)
+          setActiveRelateFile(null)
+        }
+      })
+    // only when a Project Script is clicked
+    bus &&
+      bus.take(SELECT_PROJECT_FILE, projectCypherFile => {
+        if (isStillMounted) {
+          setEditMode(true)
+          setActiveRelateFile(projectCypherFile)
+        }
+      })
+    // when a non-Project File action sets content in the editor
+    bus &&
+      bus.take(SET_CONTENT, () => {
+        if (isStillMounted) {
+          setEditMode(false)
+          setActiveRelateFile(null)
+        }
+      })
+    // Project Files is mounted
+    bus &&
+      bus.take(PROJECT_FILES_MOUNTED, () => {
+        if (isStillMounted) {
+          setProjectFilesMounted(true)
+        }
+      })
+    // Prject Files is unmounted
+    bus &&
+      bus.take(PROJECT_FILES_UNMOUNTED, () => {
+        if (isStillMounted) {
+          setProjectFilesMounted(false)
+        }
+      })
+
+    return () => {
+      isStillMounted = false
+    }
+  })
+
+  useEffect(() => {
+    // to report an error in the EditorFrame
+    bus && bus.send(PROJECT_FILE_ERROR, '')
+  }, [error])
+
+  useEffect(() => {
```

Comment:
you could look at using `useCallback()` for things like this, not sure if it suits your needs tho

Reply:
ok, Ill take a look

Reply:
Ill look into this in next PR, this could be useful